File: crates/java-omgdds/java/src/main/java/org/omg/dds/core/ReturnCode.java
// SPDX-License-Identifier: Apache-2.0
package org.omg.dds.core;

/**
 * OMG DDS Java-PSM ReturnCode — Spec §7.2.1.1.
 *
 * <p>Numeric values match the canonical DDS specification (§2.2.1.1 in DCPS).
 */
public enum ReturnCode {
    OK(0),
    ERROR(1),
    UNSUPPORTED(2),
    BAD_PARAMETER(3),
    PRECONDITION_NOT_MET(4),
    OUT_OF_RESOURCES(5),
    NOT_ENABLED(6),
    IMMUTABLE_POLICY(7),
    INCONSISTENT_POLICY(8),
    ALREADY_DELETED(9),
    TIMEOUT(10),
    NO_DATA(11),
    ILLEGAL_OPERATION(12),
    NOT_ALLOWED_BY_SECURITY(13);

    private final int code;

    ReturnCode(int code) {
        this.code = code;
    }

    public int code() {
        return code;
    }

    public boolean isOk() {
        return this == OK;
    }

    public static ReturnCode fromCode(int code) {
        for (ReturnCode rc : values()) {
            if (rc.code == code) {
                return rc;
            }
        }
        throw new IllegalArgumentException("unknown ReturnCode " + code);
    }
}

File: crates/rtc/src/return_code.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 ZeroDDS Contributors

//! `ReturnCode_t` — Spec §5.2.1.

use core::fmt;

/// `ReturnCode_t` (spec §5.2.1.1-§5.2.1.6, p. 9-10) — status codes
/// for all RTC operations that do not by nature return another
/// value.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ReturnCode {
    /// `OK` (§5.2.1.1) — Operation completed successfully.
    Ok,
    /// `ERROR` (§5.2.1.2) — Generic, unspecified failure.
    Error,
    /// `BAD_PARAMETER` (§5.2.1.3) — Illegal argument.
    BadParameter,
    /// `UNSUPPORTED` (§5.2.1.4) — Compliance-Point not implemented.
    Unsupported,
    /// `OUT_OF_RESOURCES` (§5.2.1.5) — Resource exhausted.
    OutOfResources,
    /// `PRECONDITION_NOT_MET` (§5.2.1.6) — State-Machine-Violation
    /// (e.g. `initialize` while not in the `Created` state).
    PreconditionNotMet,
}

impl ReturnCode {
    /// Spec §5.2.1.1 — `true` if `Ok`.
    #[must_use]
    pub const fn is_ok(self) -> bool {
        matches!(self, Self::Ok)
    }

    /// Converts to `Result<(), ReturnCode>` for ergonomic
    /// use of the `?` operator in the caller.
    ///
    /// # Errors
    /// Returns `Err(self)` if not `Ok`.
    pub const fn into_result(self) -> Result<(), Self> {
        match self {
            Self::Ok => Ok(()),
            other => Err(other),
        }
    }
}

impl fmt::Display for ReturnCode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Self::Ok => "OK",
            Self::Error => "ERROR",
            Self::BadParameter => "BAD_PARAMETER",
            Self::Unsupported => "UNSUPPORTED",
            Self::OutOfResources => "OUT_OF_RESOURCES",
            Self::PreconditionNotMet => "PRECONDITION_NOT_MET",
        })
    }
}

#[cfg(feature = "std")]
impl std::error::Error for ReturnCode {}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ok_is_ok_and_others_are_not() {
        // Spec §5.2.1.1.
        assert!(ReturnCode::Ok.is_ok());
        for rc in [
            ReturnCode::Error,
            ReturnCode::BadParameter,
            ReturnCode::Unsupported,
            ReturnCode::OutOfResources,
            ReturnCode::PreconditionNotMet,
        ] {
            assert!(!rc.is_ok());
        }
    }

    #[test]
    fn into_result_maps_ok_to_unit_and_others_to_err() {
        assert_eq!(ReturnCode::Ok.into_result(), Ok(()));
        assert_eq!(
            ReturnCode::PreconditionNotMet.into_result(),
            Err(ReturnCode::PreconditionNotMet)
        );
    }

    #[test]
    fn display_reports_spec_token_names() {
        assert_eq!(alloc::format!("{}", ReturnCode::Ok), "OK");
        assert_eq!(
            alloc::format!("{}", ReturnCode::PreconditionNotMet),
            "PRECONDITION_NOT_MET"
        );
        assert_eq!(
            alloc::format!("{}", ReturnCode::BadParameter),
            "BAD_PARAMETER"
        );
    }
}
